front/lib/api/assistant/citations.ts

```diff
@@ -0,0 +1,70 @@
+import _ from "lodash";
+
+import { MAX_ACTIONS_PER_STEP } from "@app/lib/api/assistant/agent";
+import { rand } from "@app/lib/utils/seeded_random";
+
+/**
+ * REFS are 3-character references used to cite documents or web pages.
+ * They are generated randomly and are unique.
+ * With this configuration, we generate 52×9×52=24336 references.
+ *
+ * Since we have currently max 8 steps, and 16 actions per step,
+ * we could have up to 190 refs per action.
+ *
+ * We currently limit to 64 refs per action.
+ */
+const REFS_PER_ACTION = 64;
+let REFS: string[] | null = null;
+const getRand = rand("chawarma");
+export const getRefs = () => {
+  const c = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".split("");
+  const nb = "123456789".split("");
+
+  if (REFS === null) {
+    REFS = c
+      .map((c1) => {
+        return nb.map((n) => {
+          return c.map((c2) => {
+            return `${c1}${n}${c2}`;
+          });
+        });
+      })
+      .flat(2);
+    // randomize
+    REFS.sort(() => {
+      const r = getRand();
+      return r > 0.5 ? 1 : -1;
+    });
+  }
+  return REFS;
+};
+
+/**
+ * Get the refs for a given action in a step.
+ */
+export const getRefsForActionInStep = ({
```

Comment:
getRefsForActionInStep is a bit of a shortcut compared to the solution that is emerging here: 
https://github.com/dust-tt/dust/blob/main/front/lib/api/assistant/actions/retrieval.ts#L268-L322

Also you will have clashes here vs retrieval actions in the same step since they don't use the same logic?

The idea of the approach in retrieval is that per step the model can make many actions of the same type and we want to load balance the topK (retrieval) or number of results (websearch) across actions of the same type to avoid overflowing the context.

Doing so, a natural way of handling citations allocation emerges.

This approach is still not perfect as we have clashes across steps which we need to fix as well, tracked here: https://github.com/dust-tt/tasks/issues/949

Can you give me 1h to harmonize the two and make it future proof for other actions that require Citations? :pray: 

This will solely impact the server side part of your PR which is minimal

Reply:
> Also you will have clashes here vs retrieval actions in the same step since they don't use the same logic?

I thought this was okay as is as as of now retrieval are still 2 characters and if this PR was approved with this logic I would have afterwards used this same logic for retrieval, and then they would be on different action index so no collision. 

> Can you give me 1h to harmonize the two and make it future proof for other actions that require Citations? 🙏

Yes sure 🙏🏻